docs/Customization/Firrtl-Transforms.rst

```diff
@@ -0,0 +1,88 @@
+Transforms are a powerful tool to take in the FIRRTL IR that is emitted from Chisel and run analysis or convert the circuit into a new form.
+
+Where to add transforms
+-----------------------
+
+In Chipyard, the FIRRTL compiler is called multiple times to create a "Top" file that contains the DUT and a "Harness" file containing the test harness, which instantiates the DUT.
+The "Harness" file does not contain the DUT's module definition or any of its submodules.
+This done by the ``tapeout`` SBT project (located in ``tools/barstools/tapeout``) which calls ``GenerateTopAndHarness`` (a function that wraps the multiple FIRRTL compiler calls and extra transforms).
+
+.. literalinclude:: ../../common.mk
+    :language: make
+    :start-after: DOC include start: FirrtlCompiler
+    :end-before: DOC include end: FirrtlCompiler
+
+If you look inside of the `tools/barstools/tapeout/src/main/scala/transforms/Generate.scala <https://github.com/ucb-bar/barstools/blob/master/tapeout/src/main/scala/transforms/Generate.scala>`__ file,
+you can see that FIRRTL is invoked twice, once for the "Top" and once for the "Harness". If you want to add transforms to just modify the DUT, you can add them to ``topTransforms``.
+Otherwise, if you want to add transforms to just modify the test harness, you can add them to ``harnessTransforms``.
+
+Examples of transforms
+----------------------
+
+There are multiple examples of transforms that you can apply and are spread across the FIRRTL ecosystem.
+Within FIRRTL there is a default set of supported transforms located in https://github.com/freechipsproject/firrtl/tree/master/src/main/scala/firrtl/transforms.
+This includes transforms that can flatten modules (``Flatten``), group modules together (``GroupAndDedup``), and more.
+
+Transforms can be standalone or can take annotations as input. Annotations are used to pass information between FIRRTL transforms. This includes information on
+what modules to flatten, group, and more. Annotations can be added to the code by
+adding them to your Chisel source or by creating a serialized annotation ``json`` file and adding it to the FIRRTL compiler
+(note: annotating the Chisel source will automatically serialize the annotation as a ``json`` snippet into the build system for you).
+**The recommended way to annotate something is to do it in the Chisel source, but not all annotation types have Chisel APIs**.
+
+Here is an example of adding a ``DontTouchAnnotation`` within the Chisel source. This annotation
+makes sure that a particular signal is not removed by the "Dead Code Elimination" pass in FIRRTL:
+
+.. code-block:: scala
+
+    class TopModule extends Module {
+        ...
+        val submod = Module(new Submodule)
+        ...
+    }
+
+    class Submodule extends Module {
+        ...
+        val some_signal := ...
+        annotate(new ChiselAnnotation {
```

Comment:
Is there no Chisel API for DontTouch? This seems like a very weird way to annotate...

Reply:
https://www.chisel-lang.org/api/SNAPSHOT/chisel3/dontTouch$.html

Reply:
I did have the ChiselAPI right underneath. I just figured that it would be clearer to show how the ChiselAnnotation is written... then show the nice Chisel wrapper for it.

Reply:
Then I think it might be worth to explicitly state in the text the most FIRRTL annotations have an equivalent Chisel annotation. However, if there doesn't exist an equivalent Chisel annotation, the way to use a FIRRTL annotation within chisel would be as follow

Reply:
You're making the Chisel wrapper seems very secondary here, while it should be the main way of using this annotation